asdf/extension/_manager.py

```diff
@@ -214,3 +225,80 @@ def get_cached_extension_manager(extensions):
 @lru_cache
 def _get_cached_extension_manager(extensions):
     return ExtensionManager(extensions)
+
+
+class ValidatorManager:
+    """
+    Wraps a list of validators and indexes them by schema property.
+
+    Parameters
+    ----------
+    validators : iterable of asdf.extension.Validator
+        List of validators to manage.
+    """
+
+    def __init__(self, validators):
+        self._validators = list(validators)
+
+        self._validators_by_schema_property = {}
+        for validator in self._validators:
+            if validator.schema_property not in self._validators_by_schema_property:
+                self._validators_by_schema_property[validator.schema_property] = set()
+            self._validators_by_schema_property[validator.schema_property].add(validator)
+
+    def validate(self, schema_property, schema_property_value, node, schema):
+        """
+        Validate an ASDF tree node against a schema property.
+
+        Parameters
+        ----------
+        schema_property : str
+            Name of the schema property (identifies the validator(s) to use).
+        schema_property_value : object
+            Value of the schema property.
+        node : asdf.tagged.Tagged
+            The ASDF node to validate.
+        schema : dict
+            The schema object that contains the property that triggered
+            the validation.
+
+        Yields
+        ------
+        asdf.exceptions.ValidationError
+        """
+        if schema_property in self._validators_by_schema_property:
+            for validator in self._validators_by_schema_property[schema_property]:
+                if _validator_matches(validator, node):
+                    yield from validator.validate(schema_property_value, node, schema)
+
+    def get_jsonschema_validators(self):
+        """
+        Get a dictionary of validator methods suitable for use
+        with the jsonschema library.
+
+        Returns
+        -------
+        dict of str: callable
+        """
+        result = {}
+
+        for schema_property in self._validators_by_schema_property:
+            result[schema_property] = self._get_jsonschema_validator(schema_property)
+
+        return result
+
+    def _get_jsonschema_validator(self, schema_property):
+        def _validator(_, schema_property_value, node, schema):
+            return self.validate(schema_property, schema_property_value, node, schema)
```

Comment:
Jsonschema appears to cache validators. The closure here captures self (ValidationManager) which I think is only holding references to the custom validators. It's probably safe to leave this as is but if we're concerned about collection of these objects we might be able to convert self to a weakref. Are there references that might warrant this change?

Reply:
Excellent! CacheInfo is super useful. I ran the suit and logged CacheInfo for all the lru_cache uses I could find (except for the legacy stuff). Here's the output (run on your source branch for this PR):
```bash
INFO     root:conftest.py:25 _get_cached_extension_manager: CacheInfo(hits=1356, misses=365, maxsize=128, currsize=128)
INFO     root:conftest.py:25 _get_cached_validator_manager: CacheInfo(hits=379, misses=4, maxsize=128, currsize=4)
INFO     root:conftest.py:25 _create_validator: CacheInfo(hits=3100, misses=10, maxsize=128, currsize=10)
INFO     root:conftest.py:25 _load_schema: CacheInfo(hits=0, misses=0, maxsize=128, currsize=0)
INFO     root:conftest.py:25 load_custom_schema: CacheInfo(hits=0, misses=0, maxsize=128, currsize=0)
INFO     root:conftest.py:25 _load_schema_cached: CacheInfo(hits=0, misses=0, maxsize=128, currsize=0)
INFO     root:conftest.py:25 _compile_uri_match_pattern: CacheInfo(hits=15555414, misses=148, maxsize=1024, currsize=148)
```
I'm not sure if the number of misses for `_get_cached_extension_manager` is reasonable given that the tests fiddle with extensions but that's outside the scope of this PR. If it looks suspicious feel free to open an issue for it (or let me know and I can open one).